crepecake/src/main/java/com/google/cloud/tools/crepecake/cache/CacheMetadata.java

```diff
@@ -29,12 +33,87 @@
 
   private final ImageLayers<CachedLayerWithMetadata> layers = new ImageLayers<>();
 
+  /** Can be used to filter layers in the metadata. */
+  static class LayerFilter {
+
+    private final ImageLayers<CachedLayerWithMetadata> layers;
+
+    @Nullable private CachedLayerType type;
+    @Nullable private Set<File> sourceFiles;
+
+    /** True if the filters are used; false otherwise. */
+    private boolean isTypeFilterEnabled = false;
```

Comment:
These fields now feel redundant. We could check if type and sourceFiles are null directly.